libr/bin/format/elf/elf.c

```diff
@@ -3546,21 +3569,18 @@ static RBinElfSymbol* Elf_(_r_bin_elf_get_symbols_imports)(ELFOBJ *bin, int type
 						ret[ret_ctr].name[0] = 0;
 					} else {
 						bool found = false;
-						j = -1;
-						while (!ret[++j].last && j < prev_ret_size) {
-							if (ret[j].offset == ret[ret_ctr].offset &&
-									strcmp (ret[j].name, "") != 0 && strcmp (ret[j].name, &strtab[st_name]) == 0
-									&& strcmp (ret[j].type, type2str (NULL, NULL, &sym[k])) == 0) {
-								found = true;
-								break;
-							}
+						RBinElfSymbol *previous = ht_pp_find (symbols_by_name, &strtab[st_name], NULL);
+						if (previous && previous->offset == ret[ret_ctr].offset &&
```

Comment:
this additional check is not enough, because if you have symbols with same name but different other properties, only one of them would be stored in the Ht, the others would be "overwritten" because they all have the same key.

Reply:
I consider this case when writing the code, but wasn't sure if it's an actual problem?

Assuming there are only 2 symbol sections (SHT_SYMTAB and SHT_DYNSYM), it would mean that first one must contain multiple symbols with identical nonempty name but different offsets or types. From what i saw identical symbols in SYMBTAB and DYNSYM are common so it's important to deduplicate those, but how likely are almost identical symbols within single table? 
If such situation does happen what's the negative effect,? Any code using the data is supposed to handle multiple symbols with identical name but different properties already due to previous deduplication logic. For human inspecting the data it would mean that in case where there are two symbols with identical name there will be 3 instead.

Reply:
look at `test/bins/elf/analysis/rust_full`, and search for `str.1`. There are many different symbols with that same name. IIRC there are other cases I saw in the past, I'll try to look for other examples, but that's already one.

Reply:
Ok, will take a look.

Reply:
Weird for some of those str symbols numbers keep increasing so they are mostly unique but low number like `str.1` get repeated. Will try to use all three properties as hashmap key like you suggested above.

Reply:
You should probably use something like:
```C

ut32 myhash(const void *k) {
  RBinElfSymbol *sym = (RBinElfSymbol *)k;
  return dosomehashwith(sym.name, sym.offset, sym.type);
}

int mycmp(const void *k1, const void *k2) {
  RBinElfSymbol *sym1 = k1, *sym2 = k2;
  return strcmp(sym1->name, sym2->name) == 0 && sym1->offset == sym2->oiffset && ...;
}

ut32 mycalcsizeK(const void *k) {
  RBinElfSymbol *sym = k;
  return strlen (k->name);
}

HtPPOptions opt = {
  .cmp = mycmp,
  .hashfn = myhash,
  .dupkey = NULL,
  .calcsizeK = mycalcsizeK,
  .freefn = myfree,
};
HtPP *ht = ht_pp_new_opt (&opt);
ht_pp_insert (ht, ret[k], ret[k]);
```

Reply:
`dupKey` should not be necessary because the `ret` objects are always valid while you use the ht (at least in this case).